A Fistful of Shotgun Shells - Nerva Edition

Cheedows opened the pull request:
Basically what it says on the tin, you can now stack shotgun shells/slugs together into a bundle of up to 4 shells. All types of shotgun shells can be stacked in this bundle, EMPs, blanks, buckshot, stun shells you name it. As part of this you get a very aesthetically pleasing dynamic sprite that will correctly render every different type of shell you have in your bundle. All credits for the lovely sprites goes to Dongwaiver.

These bundles of shells are dynamic, their sprite will change and is rendered based off of what shells you have stacked (for example, 3 slugs and 1 buckshot will be shown as 3 slugs and 1 buckshot in the item handful). Part of this is also that while they are sort of like speedloaders (in this case, shotloaders), there is no speedloader item itself. When you have used up all 4 shells or split them apart manually the shot bundle object disappears, and you are left with just shells. It doesn't leave a mess and pretty seamless to use.

Part of this update also slightly changes how shotguns load in ammo. For just the case of shotguns, you can load in either single shells one at a time like previous, or apply the handful of shells directly to the shotgun and it will load it one by one. A side effect of this is that shotholders will also be able to do this (previously shotholders could only reload double barrels, as it's considered a speedloader. Now it can also reload combat shotguns like the new handfuls, albeit only one shell at a time). Overall, the aim of this is to make shotguns more convenient to use without the hassle of managing a bunch of individual shells, and from anyone who has played Colmar, it works similar to the handful of slugs/casings that they have.

The downside of this is that these new handfuls will work similarly to the existing shotholders we have, which might make the shot holders obsolete. However, handfuls are slightly larger (normal sized instead of small like shotholders), which makes more sense, it's just a bundle of shells rather than a neatly packed array into a holder. The side effect of this balance wise is also that handfuls only fit in your backpack, while shot holders can still fit in your pocket like a magazine.

Naxodile commented on code/modules/projectiles/guns/projectile/shotgun.dm:
```diff
@@ -2,6 +2,43 @@
 /obj/item/weapon/gun/projectile/shotgun
 	icon = 'icons/urist/items/shotguns.dmi'
 
+/obj/item/weapon/gun/projectile/shotgun/load_ammo(var/obj/item/A, mob/user)
+	if(istype(A, /obj/item/ammo_magazine))
+		. = TRUE
+		var/obj/item/ammo_magazine/AM = A
+		if(caliber != AM.caliber)
+			return //incompatible
+
+		if(AM.mag_type == SPEEDLOADER)
+			if(loaded.len >= max_shells)
+				to_chat(user, "<span class='warning'>[src] is full!</span>")
+				return
+			var/obj/item/ammo_casing/C = AM.stored_ammo[AM.stored_ammo.len]
+			AM.stored_ammo-=C
+			loaded.Insert(1, C) //add to the head of the list
+			user.visible_message("[user] inserts \a [C] into [src].", "<span class='notice'>You insert \a [C] into [src].</span>")
+			playsound(loc, load_sound, 50, 1)
+			AM.update_icon()
+			if(istype(AM, /obj/item/ammo_magazine/shotbundle))
+				if(!AM.stored_ammo.len)
+					if(!user.unEquip(AM, src))
+						return
+					AM.update_icon()
+	else if(istype(A, /obj/item/ammo_casing))
+		. = TRUE
+		var/obj/item/ammo_casing/C = A
+		if(!(load_method & SINGLE_CASING) || caliber != C.caliber)
+			return //incompatible
+		if(loaded.len >= max_shells)
+			to_chat(user, "<span class='warning'>[src] is full.</span>")
+			return
+		if(!user.unEquip(C, src))
+			return
+		loaded.Insert(1, C) //add to the head of the list
+		user.visible_message("[user] inserts \a [C] into [src].", "<span class='notice'>You insert \a [C] into [src].</span>")
+		playsound(loc, load_sound, 50, 1)
+	update_icon()
+
```
```suggestion
/obj/item/weapon/gun/projectile/shotgun/load_ammo(var/obj/item/A, mob/user)
	if(istype(A, /obj/item/ammo_magazine/shotbundle))
		. = TRUE
		var/obj/item/ammo_magazine/shotbundle/SB = A
		if(caliber != SB.caliber)
			return

		if(loaded.len >= max_shells)
			to_chat(user, "<span class='warning'>[src] is full!</span>")
			return

		var/obj/item/ammo_casing/C = SB.stored_ammo[SB.stored_ammo.len]
		SB.stored_ammo -= C
		loaded.Insert(1, C)
		user.visible_message("[user] inserts \a [C] into [src].", "<span class='notice'>You insert \a [C] into [src].</span>")
		playsound(loc, load_sound, 50, 1)
		SB.update_icon()
		SB.check_ammo_count()
		return

	return ..()
```

Naxodile replied:
This removes the one-by-one for shell holders. You should honestly just mark all shotguns as accepting speedloaders and overwrite the attack_self behavior for shotguns to set allow_dump=0 save for the double barrel.

Cheedows commented on code/modules/projectiles/ammunition.dm:
```diff
@@ -143,9 +143,13 @@
 		if(!user.unEquip(C, src))
 			return
 		stored_ammo.Add(C)
+		user.visible_message("[user] insert [C] to [src].", "<span class='notice'>You insert [C] into [src].</span>")
```
Removed the message for this to cut down on spam concern.

Cheedows commented on code/modules/projectiles/ammunition/boxes.dm:
```diff
@@ -68,6 +68,64 @@
+	icon_state = "slshell"
+	caliber = "shotgun"
+	ammo_type = /obj/item/ammo_casing/shotgun
+	matter = list(MATERIAL_STEEL = 1440)
```
Changed it to be 720, which is the equivalent of 2 shells (a bundle will always have at least 2) so net neutral metal gain.

Naxodile replied:
Not all shells are the same amount, use 2 blanks and you're getting metal gain again. You're just going to need to either have it be a matter of 0 or recalc on additions/removals.

Cheedows replied:
Made the material 0, confirmed autolathes don't take it.

Cheedows commented on code/modules/projectiles/ammunition/boxes.dm:
```diff
@@ -68,6 +68,64 @@
+/obj/item/ammo_magazine/shotbundle/on_update_icon()
+	..()
+	overlays.Cut()
+	src.icon_state = "netshell"
```
Removed src references from overlays and icon_state

Naxodile commented on code/modules/projectiles/ammunition/boxes.dm:
```diff
@@ -68,6 +68,63 @@
+	overlays.Cut()
+	icon_state = "netshell"
+	var/count = 1
+	for(var/obj/item/ammo_casing/C in src.stored_ammo)
```
```suggestion
	for(var/obj/item/ammo_casing/C in stored_ammo)
```

Cheedows commented on code/modules/projectiles/ammunition/boxes.dm:
```diff
@@ -68,6 +68,64 @@
+	multiple_sprites = 1
+	w_class = ITEM_SIZE_NORMAL
+
+/obj/item/ammo_magazine/shotbundle/attack_hand(mob/user)
```
Implemented the check_ammo_count function and the mob/user param.

Cheedows commented on code/modules/projectiles/ammunition/boxes.dm:
```diff
@@ -68,6 +68,64 @@
+		..()
+		return
+
+/obj/item/ammo_magazine/shotbundle/attack_self(mob/user)
```
Same as above, implemented this change to the code.

Naxodile commented on code/modules/projectiles/ammunition/boxes.dm:
```diff
@@ -68,6 +68,44 @@
+
+/obj/item/ammo_magazine/shotbundle/on_update_icon()
+	overlays.Cut()
+	icon_state = "netshell"
```
How did I miss this

Cheedows replied:
Removed setting the dummy icon_state when updating the icon with the overlay.

Glloyd commented:
Haven't taken a look through the code yet, but move the icons for the shotbundle to a Urist .dmi please.

Naxodile reviewed (changes requested):
Invoking a squash request on this PR, all the commits are just code quality

Glloyd commented:
we can also squash it when we merge it if need be

Cheedows commented:
Oh god oh fuck, can this be reverted real quick @Glloyd @Naxodile . I was paranoid and double checked this morning with a compile, and realized that the compiler complains if you do the icon in double quotes like icon = "icons/urist/items/shotbundle.dmi". It just needs to be switched to single quotes like icon = 'icons/urist/items/shotbundle.dmi', it was from when I moved the icons to the urist dmi and switched the icon file to point to the new one.
